### ISSUE TITLE
Improve cache efficiency

### ISSUE THREAD

**Issue by replay** [raintank/metrictank#547]
Conclusions of my meeting with Dieter today:

# The current situation

Currently the chunk cache caches chunks that have been returned by Cassandra, it does not cache the absence of chunks which means if we get multiple queries for a metric/time that has no chunks in Cassandra we'll hit Cassandra for each of them.
Now we're having a case where many metrics that get queried only have very spotty data, often a metric only has 1 out of 20 potential chunks while the others are absent because there is no data. This results in a very bad caching efficiency because we need to check for each query if those absent chunks are still absent.

# Ways to improve this

There are multiple things we need to do to tackle this problem and reduce the number of queries to Cassandra

## We don't allow the insertion of data into Cassandra before already existing data

Based on this rule we know that if a given time range is queried and Cassandra only returns data in the middle or at the end of that range there won't be any older data in returned in a future query. 

Example:
- We query Cassandra for data from `600 - 3000` and Cassandra returns one chunk that's covering `2400 - 3000`.
- The data that Cassandra returned gets cached as usual, as it's happening now already
- Now we can cache an "empty cache entry" for the time range from `600 - 2400` because we know for sure that there will never be any data in Cassandra for that time range
- If we now get the same query one more time we do not need to hit Cassandra anymore because after checking the cache for the time range `600 - 3000` there is no ambiguity left

## We can take advantage of the `LastUpdate` property that we have in the memory index

First we need to make sure that on the read nodes the `LastUpdate` property is reliable, currently it can be outdated due to the `update-interval` setting which implies that the `LastUpdate` property does not always get updated right away when data is coming in. There is no reason why the read nodes need to update the index on cassandra with the metric definitions, so we should make it possible to tell the read nodes to only read the cassandra index, but never update it, only the write nodes should update it. Then we can get rid of the `update-interval` delay and always update the `LastUpdate` property when data is coming in through Kafka.

### The following is assuming that the `LastUpdate` property is reliable and up2date.

Back to the problem with the spotty data: Let's say we get queried for a given time range, but in the chunk cache we only have the first part of that range, we don't have the last piece. If we now check the `LastUpdate` time on the memory index and we see that the `LastUpdate` time matches the end of what we have in the chunk cache, then we don't need to hit Cassandra to check it for newer data.

Example:

- We receive a query that's asking for the data for `600 - 2500`
- The chunk cache has the data for `600 - 2400`
- So we check the `LastUpdate` property of the memory index, and we see the last update time is `2400`
- Now we don't need to check Cassandra for the range `2400 - 2500` because we know it won't be there

## Summary

- Introduce the concept of "empty cache entries" and use them as described
- Make it possible to disable the updating of the metric index in Cassandra, only sync from it on startup, then disable it on the read nodes
- Then make ensure that he `LastUpdate` property is updated every time data is coming in (if syncing to Cassandra is disabled)
- Then take advantage of the additional knowledge we gain from having the `LastUpdate` property

**Comment by woodsaj** [comment]
This all sounds excellent to me. I dont see any obvious issues in the suggested changes and to me it looks like it will solve most of the caching inefficiencies we are currently seeing.

**Comment by Dieterbe** [comment]
ditto. LGTM
also as discussed there's a few edge cases in which the assumption "We don't allow the insertion of data into Cassandra before already existing data" is not true: importing data and rewrite / overwrite (#511) but neither is a problem. when importing, we could just restart MT, and for 511 we can just invalidate the cache of series being rewritten.

**Comment by woodsaj** [comment]
>  when importing, we could just restart MT,

@replay can we add an API method to clear the cache?  It would also be really handy for resolving any cache issues if pop up, without having to restart all of the MT instances.

**Comment by replay** [comment]
@woodsaj yeah, i guess that's a good idea. We'll probably want to have it anyway in these two cases that Dieter just mentioned ^^

**Comment by Dieterbe** [comment]
It would be great to make some progress on the "accurate lastupdate" stuff, then we can remove the 24h safety buffer from the index filtering logic.  It has interfered with a few customers, e.g. when they ramp up resolution of data, they won't be able to enjoy the higher res until the old data hasn't been written to for 24h longer than what's needed.

**Comment by replay** [comment]
@dieterbe does that look right? https://github.com/raintank/metrictank/pull/565/files
I think with this patch we could set `update-cassandra-index = false` and `update-interval = 0` on the read nodes, then the `LastUpdate` should always be up2date without delay.
